### PR TITLE
[Distribution] Implement sharding support for Keras PyDataset adapter

### PR DESCRIPTION
**Pull request by buildwithsuhana**
## Description
This PR implements native sharding support for keras.utils.PyDataset (and its underlying enqueuer), enabling parallel Python-based data pipelines in multi-process distributed training.

   * `PyDatasetAdapter` Sharding: Implements logic to shard indices across model replicas. Each process now only iterates over its designated data shard, significantly reducing redundant I/O and preprocessing in a distributed cluster.
   
[4th PR]: This PR should be rebased after merging of PR https://github.com/keras-team/keras/pull/23043

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/trainers/data_adapters/py_dataset_adapter.py [hunk 1]
```diff
@@ -204,12 +204,24 @@ def __init__(
         class_weight=None,
         shuffle=False,
     ):
+        from keras.src.distribution import distribution_lib
+
```
Comment (hertschuh) [inline keras/src/trainers/data_adapters/py_dataset_adapter.py]: This can't be at the top?

File: keras/src/trainers/data_adapters/py_dataset_adapter.py [hunk 2]
```diff
@@ -251,29 +265,33 @@ def _standardize_batch(self, batch):
...
-        indices = range(self.py_dataset.num_batches)
+        num_batches = self.py_dataset.num_batches
+        indices = list(range(num_batches))
         if self.shuffle:
-            indices = list(indices)
-            random.shuffle(indices)
+            if self._num_data_shards > 1:
+                random.Random(self._epoch).shuffle(indices)
+            else:
+                random.shuffle(indices)
```
Comment (hertschuh) [inline keras/src/trainers/data_adapters/py_dataset_adapter.py]: Do we need to make the distinction between the 2 cases? Can we use `random.Random(self._epoch).shuffle(indices)` always? Or is it for backwards compatibility?

File: keras/src/trainers/data_adapters/py_dataset_adapter.py [hunk 3]
```diff
@@ -473,7 +499,7 @@ def is_running(self):
         """
         return self.running
 
-    def start(self):
+    def start(self, epoch=None):
```
Comment (hertschuh) [inline keras/src/trainers/data_adapters/py_dataset_adapter.py]: nitpick: remove the `=None` to force all use cases to pass an `epoch`.

File: keras/src/trainers/data_adapters/py_dataset_adapter.py [hunk 4]
```diff
@@ -622,7 +655,20 @@ def _run(self):
                 indices = range(self.py_dataset.num_batches)
                 if self.shuffle:
                     indices = list(indices)
-                    random.shuffle(indices)
+                    if self._num_data_shards > 1:
+                        random.Random(self.epoch).shuffle(indices)
+                    else:
+                        random.shuffle(indices)
```
Comment (hertschuh) [inline keras/src/trainers/data_adapters/py_dataset_adapter.py]: Same question about simplifying here.

File: keras/src/trainers/data_adapters/py_dataset_adapter_test.py [hunk 5]
```diff
@@ -453,3 +454,149 @@ def __getitem__(self, idx):
...
+        )
+        try:
+            self.assertFalse(enqueuer.is_running())
+            enqueuer.start()
+            self.assertTrue(enqueuer.is_running())
+            enqueuer.start()
+            self.assertTrue(enqueuer.is_running())
+        finally:
+            enqueuer.stop()
+        self.assertFalse(enqueuer.is_running())
```
Comment (hertschuh) [inline keras/src/trainers/data_adapters/py_dataset_adapter_test.py]: I'm not sure I follow what this is testing that is not covered indirectly by the other tests.

File: keras/src/trainers/data_adapters/py_dataset_adapter.py (hunk 1 above)
Comment (buildwithsuhana) [reply keras/src/trainers/data_adapters/py_dataset_adapter.py]: Moved to top

File: keras/src/trainers/data_adapters/py_dataset_adapter.py (hunk 2 above)
Comment (buildwithsuhana) [reply keras/src/trainers/data_adapters/py_dataset_adapter.py]: Done

File: keras/src/trainers/data_adapters/py_dataset_adapter.py (hunk 3 above)
Comment (buildwithsuhana) [reply keras/src/trainers/data_adapters/py_dataset_adapter.py]: Done

File: keras/src/trainers/data_adapters/py_dataset_adapter.py (hunk 4 above)
Comment (buildwithsuhana) [reply keras/src/trainers/data_adapters/py_dataset_adapter.py]: Simplified

File: keras/src/trainers/data_adapters/py_dataset_adapter_test.py (hunk 5 above)
Comment (buildwithsuhana) [reply keras/src/trainers/data_adapters/py_dataset_adapter_test.py]: removed

**Review by hertschuh — APPROVED** [review]
Thanks!
